### PR TITLE
Added support for crawling grants and applying Hive Metastore UDF ACLs

### LINKED ISSUE #808: [FEATURE]: Crawl and Apply grants for SQL UDFs
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem statement

SQL UDFs grants are currently not crawled and applied during groups migration. 

### Proposed Solution

Add sql udf crawler and update existing grants module.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by mwojtyczka**
## Changes
Added support for crawling grants and applying Hive Metastore UDF ACLs.

Resolves #808 

### Functionality 

- [x] added new crawler for udfs
- [x] added handling of udfs to the existing TableAclSupport
- [x] added new column to inventory permissions table for udfs

### Tests

- [x] tested manually
- [x] added unit tests
- [x] added integration tests

### CODE REVIEW

File: src/databricks/labs/ucx/hive_metastore/udfs.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        yield from self._fetch("SHOW DATABASES")
+
+    def snapshot(self) -> list[Udf]:
+        """
+        Takes a snapshot of tables in the specified catalog and database.
+
+        Returns:
+            list[Table]: A list of Udf objects representing the snapshot of tables.
+        """
+        return self._snapshot(partial(self._try_load), partial(self._crawl))
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/udfs.py:53]: You don't need partial, as no arguments are passed down to tryload/crawl

File: src/databricks/labs/ucx/hive_metastore/udfs.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    def _all_databases(self) -> Iterator[Row]:
+        yield from self._fetch("SHOW DATABASES")
+
+    def snapshot(self) -> list[Udf]:
+        """
+        Takes a snapshot of tables in the specified catalog and database.
+
+        Returns:
+            list[Table]: A list of Udf objects representing the snapshot of tables.
```
Comment (nfx) [inline src/databricks/labs/ucx/hive_metastore/udfs.py:51]:
```suggestion
            list[Udf]: A list of Udf objects representing the snapshot of tables.
```

File: src/databricks/labs/ucx/mixins/fixtures.py [hunk 3]
```diff
@@ -1023,6 +1024,47 @@ def remove(table_info: TableInfo):
...
+        logger.info(f"Function {udf_info.full_name} crated")
+        return udf_info
+
+    def remove(udf_info: FunctionInfo):
+        try:
+            sql_backend.execute(f"DROP FUNCTION IF EXISTS {udf_info.full_name}")
+        except RuntimeError as e:
+            if "Cannot drop udf " in str(e):
+                sql_backend.execute(f"DROP FUNCTION IF EXISTS {udf_info.full_name}")
+            elif "SCHEMA_NOT_FOUND" in str(e):
```
Comment (nfx) [inline src/databricks/labs/ucx/mixins/fixtures.py:1060]: Replace it with "except NotFound" - we're already mapping that exception

File: src/databricks/labs/ucx/mixins/fixtures.py [hunk 4]
```diff
@@ -1023,6 +1024,47 @@ def remove(table_info: TableInfo):
...
+        )
+
+        logger.info(f"Function {udf_info.full_name} crated")
+        return udf_info
+
+    def remove(udf_info: FunctionInfo):
+        try:
+            sql_backend.execute(f"DROP FUNCTION IF EXISTS {udf_info.full_name}")
+        except RuntimeError as e:
+            if "Cannot drop udf " in str(e):
```
Comment (nfx) [inline src/databricks/labs/ucx/mixins/fixtures.py:1058]: Why that is happening?...

File: tests/integration/hive_metastore/test_grants.py [hunk 5]
```diff
@@ -53,3 +55,27 @@ def test_all_grants_in_databases(sql_backend, inventory_schema, make_schema, mak
     assert all_grants[f"{group_b.display_name}.{view_c.full_name}"] == "MODIFY"
     assert all_grants[f"{group_b.display_name}.{view_d.full_name}"] == "MODIFY"
     assert all_grants[f"{group_b.display_name}.{table_e.full_name}"] == "MODIFY"
+
+
+@retried(on=[NotFound, TimeoutError], timeout=timedelta(minutes=3))
```
Comment (nfx) [inline tests/integration/hive_metastore/test_grants.py:60]: Remove TimeoutError

**Review by nfx — APPROVED** [review]
Overall lgtm, couple of nits though

File: src/databricks/labs/ucx/hive_metastore/udfs.py (hunk 1 above)
Comment (mwojtyczka) [reply src/databricks/labs/ucx/hive_metastore/udfs.py:53]: corrected

File: src/databricks/labs/ucx/mixins/fixtures.py (hunk 3 above)
Comment (mwojtyczka) [reply src/databricks/labs/ucx/mixins/fixtures.py:1060]: corrected

File: src/databricks/labs/ucx/mixins/fixtures.py (hunk 4 above)
Comment (mwojtyczka) [reply src/databricks/labs/ucx/mixins/fixtures.py:1058]: i removed this. I haven't seen that happening.

File: tests/integration/hive_metastore/test_grants.py (hunk 5 above)
Comment (mwojtyczka) [reply tests/integration/hive_metastore/test_grants.py:60]: removed

**Review by nfx — APPROVED** [review]
lgtm

**Pull request merged**